src/ucp/rndv/rndv_am.c

```diff
@@ -135,3 +135,96 @@ ucp_proto_t ucp_rndv_am_bcopy_proto = {
     .abort    = ucp_proto_rndv_am_bcopy_abort,
     .reset    = ucp_proto_reset_fatal_not_implemented
 };
+
+static UCS_F_ALWAYS_INLINE ucs_status_t ucp_rndv_am_zcopy_send_func(
+        ucp_request_t *req, const ucp_proto_multi_lane_priv_t *lpriv,
+        ucp_datatype_iter_t *next_iter, ucp_lane_index_t *lane_shift)
+{
+    const size_t hdr_size      = sizeof(ucp_request_data_hdr_t);
+    const size_t max_payload   = ucp_proto_multi_max_payload(req, lpriv,
+                                                             hdr_size);
+    ucp_request_data_hdr_t hdr = {
+        .req_id = req->send.rndv.remote_req_id,
+        .offset = req->send.state.dt_iter.offset
+    };
+    uct_iov_t iov[UCP_MAX_IOV];
+    size_t iov_count;
+
+    ucs_assert(lpriv->super.max_iov > 1);
+    iov_count = ucp_datatype_iter_next_iov(&req->send.state.dt_iter,
+                                           max_payload, lpriv->super.md_index,
+                                           UCP_DT_MASK_CONTIG_IOV, next_iter,
+                                           iov, lpriv->super.max_iov - 1);
+
+    return uct_ep_am_zcopy(ucp_ep_get_lane(req->send.ep, lpriv->super.lane),
+                           UCP_AM_ID_RNDV_DATA, &hdr, hdr_size, iov, iov_count,
+                           0, &req->send.state.uct_comp);
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_rndv_am_zcopy_complete(ucp_request_t *req)
+{
+    if (req->send.rndv.rkey != NULL) {
+        ucp_proto_rndv_rkey_destroy(req);
+    }
+
+    return ucp_request_invoke_uct_completion_success(req);
+}
+
+static ucs_status_t ucp_proto_rndv_am_zcopy_progress(uct_pending_req_t *uct_req)
+{
+    ucp_request_t *req = ucs_container_of(uct_req, ucp_request_t, send.uct);
+
+    /* coverity[tainted_data_downcast] */
+    return ucp_proto_multi_zcopy_progress(req, req->send.proto_config->priv,
+                                          NULL, UCT_MD_MEM_ACCESS_LOCAL_READ,
+                                          UCP_DT_MASK_CONTIG_IOV,
+                                          ucp_rndv_am_zcopy_send_func,
+                                          ucp_rndv_am_zcopy_complete,
+                                          ucp_proto_request_zcopy_completion);
+}
+
+static ucs_status_t
+ucp_proto_rndv_am_zcopy_init(const ucp_proto_init_params_t *init_params)
+{
+    ucp_context_t *context               = init_params->worker->context;
+    ucp_proto_multi_init_params_t params = {
+        .super.super         = *init_params,
+        .super.cfg_thresh    = context->config.ext.zcopy_thresh,
```

Comment:
I think it is better to have a gtest which is exactly uses this protocol. Since there is a configurable threshold.